### Import Libraries
This cell imports the necessary libraries for data manipulation, numerical operations, and machine learning preprocessing.

In [14]:
import pandas as pd  # Import the pandas library for data manipulation
import numpy as np  # Import the numpy library for numerical operations
from sklearn.preprocessing import MinMaxScaler  # Import MinMaxScaler for feature scaling

from warnings import filterwarnings  # Import filterwarnings to manage warnings
filterwarnings('ignore')  # Ignore warnings to keep the output clean

### Define Data and Create DataFrame
This cell defines the raw data for drivers, their teams, performance metrics, and historical Canadian GP statistics. It then creates a Pandas DataFrame from this data.

In [15]:
data = {
    'Driver': ['Antonelli', 'Russell', 'Norris', 'Piastri', 'Verstappen',
               'Leclerc', 'Hamilton', 'Sainz', 'Alonso', 'Stroll'],
    'Team': ['Mercedes', 'Mercedes', 'McLaren', 'McLaren', 'Red Bull',
             'Ferrari', 'Ferrari', 'Williams', 'Aston Martin', 'Aston Martin'],
    'Season_Points': [72, 63, 54, 48, 35, 30, 28, 22, 15, 8], # Current season points
    'Canadian_GP_Wins': [0, 1, 0, 0, 1, 0, 7, 0, 2, 0], # Historical Canadian GP wins
    'Canadian_GP_Podiums': [0, 1, 1, 0, 2, 1, 9, 0, 3, 0], # Historical Canadian GP podiums
    'Avg_Finish_2026': [2.1, 2.8, 3.2, 3.9, 5.1, 5.8, 6.0, 7.2, 8.5, 10.1], # Average finish position in 2026
    'DNF_Count_2026': [0, 0, 1, 0, 1, 1, 0, 1, 0, 1] # Did Not Finish (DNF) count in 2026
}

df = pd.DataFrame(data) # Create a Pandas DataFrame from the dictionary

### Display Initial DataFrame
This cell displays the first few rows of the created DataFrame to show its structure and content.

In [16]:
df # Display the entire DataFrame to review its contents

,Driver,Team,Season_Points,Canadian_GP_Wins,Canadian_GP_Podiums,Avg_Finish_2026,DNF_Count_2026
0,Antonelli,Mercedes,72,0,0,2.1,0
1,Russell,Mercedes,63,1,1,2.8,0
2,Norris,McLaren,54,0,1,3.2,1
3,Piastri,McLaren,48,0,0,3.9,0
4,Verstappen,Red Bull,35,1,2,5.1,1
5,Leclerc,Ferrari,30,0,1,5.8,1
6,Hamilton,Ferrari,28,7,9,6.0,0
7,Sainz,Williams,22,0,0,7.2,1
8,Alonso,Aston Martin,15,2,3,8.5,0
9,Stroll,Aston Martin,8,0,0,10.1,1


### Calculate Canadian Podium Rate
This cell calculates a 'Canadian_Podium_Rate' by dividing the 'Canadian_GP_Podiums' by an estimated number of races (10), providing a normalized historical performance metric for the Canadian GP.

In [17]:
# Calculate the historical podium rate at the Canadian GP
df['Canadian_Podium_Rate'] = df['Canadian_GP_Podiums'] / 10  # Divide by ~10 races sampled for normalization

df # Display the DataFrame with the new 'Canadian_Podium_Rate' column

,Driver,Team,Season_Points,Canadian_GP_Wins,Canadian_GP_Podiums,Avg_Finish_2026,DNF_Count_2026,Canadian_Podium_Rate
0,Antonelli,Mercedes,72,0,0,2.1,0,0.0
1,Russell,Mercedes,63,1,1,2.8,0,0.1
2,Norris,McLaren,54,0,1,3.2,1,0.1
3,Piastri,McLaren,48,0,0,3.9,0,0.0
4,Verstappen,Red Bull,35,1,2,5.1,1,0.2
5,Leclerc,Ferrari,30,0,1,5.8,1,0.1
6,Hamilton,Ferrari,28,7,9,6.0,0,0.9
7,Sainz,Williams,22,0,0,7.2,1,0.0
8,Alonso,Aston Martin,15,2,3,8.5,0,0.3
9,Stroll,Aston Martin,8,0,0,10.1,1,0.0


### Calculate Reliability Score
This cell assigns a 'Reliability_Score' based on the 'DNF_Count_2026'. Drivers with no DNFs get a score of 1.0, one DNF gets 0.7, and more than one DNF gets 0.4.

In [18]:
# Assign a 'Reliability_Score' based on 'DNF_Count_2026'
# 1.0 for 0 DNFs, 0.7 for 1 DNF, and 0.4 for more than 1 DNF
df['Reliability_Score'] = np.where(df['DNF_Count_2026'] == 0, 1.0,
                                   np.where(df['DNF_Count_2026'] == 1, 0.7, 0.4))

### Display DataFrame with Reliability Score
This cell displays the DataFrame after the 'Reliability_Score' has been added.

In [19]:
df # Display the DataFrame to show the newly added 'Reliability_Score' column

,Driver,Team,Season_Points,Canadian_GP_Wins,Canadian_GP_Podiums,Avg_Finish_2026,DNF_Count_2026,Canadian_Podium_Rate,Reliability_Score
0,Antonelli,Mercedes,72,0,0,2.1,0,0.0,1.0
1,Russell,Mercedes,63,1,1,2.8,0,0.1,1.0
2,Norris,McLaren,54,0,1,3.2,1,0.1,0.7
3,Piastri,McLaren,48,0,0,3.9,0,0.0,1.0
4,Verstappen,Red Bull,35,1,2,5.1,1,0.2,0.7
5,Leclerc,Ferrari,30,0,1,5.8,1,0.1,0.7
6,Hamilton,Ferrari,28,7,9,6.0,0,0.9,1.0
7,Sainz,Williams,22,0,0,7.2,1,0.0,0.7
8,Alonso,Aston Martin,15,2,3,8.5,0,0.3,1.0
9,Stroll,Aston Martin,8,0,0,10.1,1,0.0,0.7


### Calculate Inverse Average Finish
This cell calculates 'Inverse_Avg_Finish' as the reciprocal of 'Avg_Finish_2026'. A lower average finish position indicates better performance, so taking the inverse makes higher values better for scoring.

In [20]:
# Calculate 'Inverse_Avg_Finish': a higher value indicates a better average finish position
df['Inverse_Avg_Finish'] = 1 / df['Avg_Finish_2026']

df # Display the DataFrame with the new 'Inverse_Avg_Finish' column

,Driver,Team,Season_Points,Canadian_GP_Wins,Canadian_GP_Podiums,Avg_Finish_2026,DNF_Count_2026,Canadian_Podium_Rate,Reliability_Score,Inverse_Avg_Finish
0,Antonelli,Mercedes,72,0,0,2.1,0,0.0,1.0,0.476190
1,Russell,Mercedes,63,1,1,2.8,0,0.1,1.0,0.357143
2,Norris,McLaren,54,0,1,3.2,1,0.1,0.7,0.312500
3,Piastri,McLaren,48,0,0,3.9,0,0.0,1.0,0.256410
4,Verstappen,Red Bull,35,1,2,5.1,1,0.2,0.7,0.196078
5,Leclerc,Ferrari,30,0,1,5.8,1,0.1,0.7,0.172414
6,Hamilton,Ferrari,28,7,9,6.0,0,0.9,1.0,0.166667
7,Sainz,Williams,22,0,0,7.2,1,0.0,0.7,0.138889
8,Alonso,Aston Martin,15,2,3,8.5,0,0.3,1.0,0.117647
9,Stroll,Aston Martin,8,0,0,10.1,1,0.0,0.7,0.099010


### Define Features for Scaling
This cell defines the list of features that will be used for calculating the 'Podium_Score' and creates a copy of the DataFrame to store scaled values.

In [21]:
features = ['Season_Points', 'Canadian_Podium_Rate', 'Reliability_Score', 'Inverse_Avg_Finish'] # Define features for scoring

df_scaled = df.copy() # Create a copy of the DataFrame to store scaled feature values

### Scale Features
This cell initializes a `MinMaxScaler` and applies it to the selected features in `df_scaled` to normalize their values between 0 and 1. This ensures all features contribute equally to the 'Podium_Score'.

In [22]:
scaler = MinMaxScaler() # Initialize the MinMaxScaler

# Apply Min-Max scaling to the selected features, transforming values to a 0-1 range
df_scaled[features] = scaler.fit_transform(df[features])

### Calculate Podium Score
This cell calculates the 'Podium_Score' by taking a weighted sum of the scaled features. The weights are assigned based on the perceived importance of each feature for predicting podium potential.

In [23]:
weights = np.array([0.40, 0.25, 0.20, 0.15]) # Define weights for each feature (Season_Points, Canadian_Podium_Rate, Reliability_Score, Inverse_Avg_Finish)
# Calculate 'Podium_Score' by taking a weighted sum of the scaled features
df['Podium_Score'] = df_scaled[features].values @ weights

### Sort and Rank Drivers
This cell sorts the DataFrame by 'Podium_Score' in descending order and assigns a 'Predicted_Rank' to each driver. It then displays a subset of columns for the ranked drivers.

In [24]:
df_sorted = df.sort_values('Podium_Score', ascending=False).reset_index(drop=True) # Sort DataFrame by 'Podium_Score' in descending order and reset index
df_sorted['Predicted_Rank'] = df_sorted.index + 1 # Assign a 'Predicted_Rank' based on the sorted order

# Display relevant columns for the ranked drivers
df_sorted[['Driver', 'Team', 'Season_Points', 'Podium_Score', 'Predicted_Rank']]

,Driver,Team,Season_Points,Podium_Score,Predicted_Rank
0,Antonelli,Mercedes,72,0.750000,1
1,Russell,Mercedes,63,0.674184,2
2,Hamilton,Ferrari,28,0.601906,3
3,Piastri,McLaren,48,0.512596,4
4,Norris,McLaren,54,0.400180,5
5,Alonso,Aston Martin,15,0.334495,6
6,Verstappen,Red Bull,35,0.262908,7
7,Leclerc,Ferrari,30,0.194470,8
8,Sainz,Williams,22,0.103359,9
9,Stroll,Aston Martin,8,0.000000,10


### Display Top 5 Contenders
This cell extracts the top 5 drivers based on their 'Podium_Score' and prints them, highlighting the predicted podium contenders for the Canadian GP.

In [25]:
top5 = df_sorted.head(5) # Get the top 5 drivers from the sorted DataFrame
print("\n🏆 Top 5 Canadian GP Podium Contenders:") # Print a header for the top contenders
# Print the predicted rank, driver, team, and podium score for the top 5, without the DataFrame index
print(top5[['Predicted_Rank', 'Driver', 'Team', 'Podium_Score']].to_string(index=False))


🏆 Top 5 Canadian GP Podium Contenders:
 Predicted_Rank    Driver     Team  Podium_Score
              1 Antonelli Mercedes      0.750000
              2   Russell Mercedes      0.674184
              3  Hamilton  Ferrari      0.601906
              4   Piastri  McLaren      0.512596
              5    Norris  McLaren      0.400180


### Final Analysis Result
This notebook performed a predictive analysis to identify the top 5 podium contenders for the Canadian Grand Prix based on various performance metrics, historical data, and reliability. By calculating a 'Podium_Score' from weighted and scaled features, we have ranked the drivers and identified the most promising candidates.

**Key Findings:**
Our analysis predicts the following drivers as the top contenders for a podium finish at the Canadian GP:

| Predicted_Rank | Driver    | Team     | Podium_Score |
|----------------|-----------|----------|--------------|
| 1              | Antonelli | Mercedes | 0.750000     |
| 2              | Russell   | Mercedes | 0.674184     |
| 3              | Hamilton  | Ferrari  | 0.601906     |
| 4              | Piastri   | McLaren  | 0.512596     |
| 5              | Norris    | McLaren  | 0.400180     |

This ranking suggests Mercedes and Ferrari drivers have strong potential, with McLaren also having a contender in the top 5.